File: 146. LRU Cache/LRUCache.py
class Node:
    def __init__(self, key = None, val = None):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None

class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.hashmap = {}
        self.head = Node()
        self.tail = Node()
        self.head.next = self.tail
        self.tail.prev = self.head

    def move_node_to_head(self, key):
        curr = self.hashmap[key]
        curr.next.prev = curr.prev
        curr.prev.next = curr.next
        curr.prev = self.head
        curr.next = self.head.next
        self.head.next.prev = curr
        self.head.next = curr


    def get(self, key: int) -> int:
        if self.hashmap.__contains__(key):
            self.move_node_to_head(key)
        res = self.hashmap.get(key, -1)
        if res == -1:
            return res
        else:
            return res.val

    def put(self, key: int, value: int) -> None:
        if self.hashmap.__contains__(key):
            self.hashmap[key].val = value
            self.move_node_to_head(key)
        else:
            if len(self.hashmap) == self.capacity:
                self.hashmap.pop(self.tail.prev.key)
                self.tail.prev.prev.next = self.tail
                self.tail.prev = self.tail.prev.prev
        
            curr = Node(key, value)
            self.hashmap[key] = curr
            curr.prev = self.head
            curr.next = self.head.next
            self.head.next.prev = curr
            self.head.next = curr           



# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 146. LRU Cache/LRUCache.java
class Node {
    int key;
    int val;
    Node prev;
    Node next;

    public Node() {

    }

    public Node(int key, int val) {
        this.key = key;
        this.val = val;
        this.prev = null;
        this.next = null;
    }
}

class LRUCache {
    int capacity;
    HashMap<Integer, Node> map;
    Node head;
    Node tail;
    public LRUCache(int capacity) {
        this.capacity = capacity;
        map = new HashMap<>();
        head = new Node();
        tail = new Node();
        head.next = tail;
        tail.prev = head;
    }

    public void move_node_to_head(int key) {
        Node curr = map.get(key);
        curr.prev.next = curr.next;
        curr.next.prev = curr.prev;
        
        curr.next = head.next;
        curr.prev = head;
        head.next.prev = curr;
        head.next = curr;
    }
    
    public int get(int key) {
        if (map.containsKey(key)) {
            move_node_to_head(key);
        }
        Node res = new Node(-1, -1);
        res = map.getOrDefault(key, res);
        return res.val;
    }
    
    public void put(int key, int value) {
        Node curr = new Node(key, value);
        if (map.containsKey(key)) {
            curr = map.get(key);
            curr.val = value;
            map.put(key, curr);
            move_node_to_head(key);
        } else {
            if (map.size() == capacity) {
                map.remove(tail.prev.key);
                tail.prev.prev.next = tail;
                tail.prev = tail.prev.prev;
            }
            map.put(key, curr);
            curr.prev = head;
            curr.next = head.next;
            head.next.prev = curr;
            head.next = curr;
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */